### DISCUSSION TITLE
User not found message information not displayed when reset password

### DISCUSSION THREAD

**Discussion by klauswr** [keycloak/keycloak#28966] [Q&A]
I would like to inform the user about a wrong userId entered in reset password dialog. 

In Stackoverflow there was the same question: (https://stackoverflow.com/questions/48327141/keycloack-user-not-found-message-information-not-displayed-when-reset-password):

> On my keycloak server, when i go to page "reset password" and i enter a bad username, the return message say "You should receive an email shortly with further instructions." But, i want to send a message like "username not found"

The answer there was: 

> Forgot password has default implementation. If you want to change it you have to: 1. create your own provider for "Choose User"; 2. create a new FLOW with it. 3. change bidings for Reset Credentials with new FLOW. More information you can find on: http://www.keycloak.org/docs/latest/server_development/index.html

I am using keycloak in a docker container: **_image: quay.io/keycloak/keycloak:21.0_** and I am not aware of such a thing like create your own provider for "Choose User".

Anyway, what I tried was straight forward: 
I added a step "**User Validation**" after "**Choose User**"
![image](https://github.com/keycloak/keycloak/assets/4521155/21cfadf8-8e9c-4245-b25f-ffa1872b9f0f)

In principle that did the trick, but: If a user enters an unknown userId he gets this result in the browser: 
`{
error: "invalid_grant",
error_description: "Invalid user credentials"
}`

Is there any way to change this to a message within the dialog itself?

**Comment by codespearhead** [comment]
This is by design, as [the Open Web Application Security Project (OWAPS) recommends](https://cheatsheetseries.owasp.org/cheatsheets/Forgot_Password_Cheat_Sheet.html#forgot-password-request):

> Return a consistent message for both existent and non-existent accounts.

Allowing user enumeration through the password reset feature is a [bad security practice](https://owasp.org/www-project-web-security-testing-guide/v42/4-Web_Application_Security_Testing/03-Identity_Management_Testing/04-Testing_for_Account_Enumeration_and_Guessable_User_Account).

**Reply by klauswr** [reply]
So even though: how can I achieve the solution sketched before **anyway**?

**Reply by codespearhead** [reply]
If you want to make the most important component of your infrastructure, which is the authentication and authorization server, more vulnerable to cyber attacks for learning purposes, you'll have to create a Conditional Authenticator (see #29044 ).

Related issues: #29059 and #26293 .
